internal/registry/registry.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2021 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registry
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+
+	"github.com/tetratelabs/car/internal"
+	"github.com/tetratelabs/car/internal/httpclient"
+	"github.com/tetratelabs/car/internal/registry/docker"
+	"github.com/tetratelabs/car/internal/registry/github"
+)
+
+type registry struct {
+	host, path, baseURL string
+	roundTripper        http.RoundTripper
+}
+
+type contextRoundTripper struct {
+}
+
+func (f *contextRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return httpclient.TransportFromContext(req.Context()).RoundTrip(req)
+}
+
+// New returns a new instance of a remote registry
+func New(ref reference.Named) internal.Registry {
+	domain := reference.Domain(ref)
+	path := reference.Path(ref)
+	host := domain
+	var roundTripper http.RoundTripper
+	switch domain {
+	case "docker.io":
+		host = "index.docker.io"
+		roundTripper = docker.NewRoundTripper(path)
+	case "ghcr.io":
+		roundTripper = github.NewRoundTripper()
+	default:
+		roundTripper = &contextRoundTripper{}
+	}
+	baseURL := fmt.Sprintf("https://%s/v2/%s", host, path)
+	return &registry{host: host, path: path, baseURL: baseURL, roundTripper: roundTripper}
```

Comment:
looks like roundTripper is only used via `r.httpClient()` so how about creating `httpclient.HTTPClient` here and have it as a field member of the registry struct?
```suggestion
	if transport == nil {
		transport = httpclient.TransportFromContext(ctx)
	}
	return &registry{host: host, path: path, baseURL: baseURL, httpClient: httpclient.New(transport)}
```

Reply:
I'll try this and see where it goes. Before we used a couple things, but that was when the code had more scope (ex also checking github releases page)

Reply:
yeah we probably can do this because we have a context already (the one from urfave).. will try it out